Refactor out a utility to get the top level packages given a package graph.

nkolev92 opened the pull request:
## Bug

Fixes: https://github.com/NuGet/Home/issues/8316
Regression: No  
* Last working version:   
* How are we preventing it in future:   

## Fix

Details: Refactor out a couple of utility methods helpful for discovering top level packages given a package graph. 

Any naming suggestions are appreciated. I'll think about it too, might end up changing some before it's all said and done.

## Testing/Validation

Tests Added: Yes 
Reason for not adding tests:  
Validation:

nkolev92 committed:
refactoring attempt 1

nkolev92 committed:
finish up refactoring

nkolev92 committed:
work

nkolev92 committed:
add tests

loic-sharma commented on src/NuGet.Core/NuGet.PackageManagement/IPackageWithDependants.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace NuGet.PackageManagement
+{
+    public interface IPackageWithDependants
```
```suggestion
    public interface IPackageWithDependents
```

Edit: I looked this up - "dependants" is an older and less common spelling, but it's valid! Feel free to keep it as-is if you prefer it

nkolev92 replied:
🙃 
I like dependants as a noun usually. 

I don't mind changing if others think so as well.

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/IPackageWithDependants.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        PackageIdentity Identity { get; }
+
+        IList<PackageIdentity> DependingPackages { get; }
```
Depending is a verb. Dependant is a noun. Properties and variable names should be nouns 😄

nkolev92 replied:
good catch.

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs:
```diff
@@ -1296,11 +1296,25 @@ private static bool IsPackageReferenceAutoReferenced(PackageReference package)
         }
 
+        /// <summary>
+        /// The package dependency info for the given project. The project needs to be packages.config, otherwise returns an empty list.
```
> The project needs to be packages.config, otherwise returns an empty list

Why?

If you remove the ` if (nuGetProject is MSBuildNuGetProject)` and keep the old method body, will it work for all projects?

nkolev92 replied:
nope, it won't. It would've thrown. I also don't see a reason to make it work retroactively yet. So this just makes it more explicit. It's only used in the migrator right now.

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,123 @@
+    public static class PackageGraphAnalysisUtilities
+    {
+        public static Func<IPackageWithDependants, bool> TopLevelPackagesPredicate = dependencyItem => !dependencyItem.DependingPackages.Any();
+        public static Func<IPackageWithDependants, bool> TransitivePackagesPredicate = dependencyItem => dependencyItem.DependingPackages.Any();
```
why not make these two "proper" methods?

nkolev92 replied:
Can you be more specific?

zivkan replied:
```csharp
public static bool TransitivePackagesPredicate(IPackageWithDependants dependencyItem)
{
  return dependencyItem.DependingPackages.Any();
}
```

or 
```csharp
public static bool TransitivePackagesPredicate(IPackageWithDependants dependencyItem) => return dependencyItem.DependingPackages.Any();
```

Not that it's a big performance impact, but currently it's a property that returns a Func<>, so a new Func<> is probably allocated on the heap and returned every call, and when the caller wants to call it, it has to go through an extra layer on indirection, as Func<> is kind of like the C# equivalent of a function pointer in C. The runtime needs to dereference the function address before it can be called.

Func<> and Action<> are awsome for functional programming, so used as input parameters in methods, or as a kind of polymorphic programming tool, but it offers nothing as an alternative to class methods.

nkolev92 replied:
That's fine. I can do that.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/NuGetProjectUpgradeWindowModel.cs:
```diff
@@ -158,29 +157,13 @@ private void PromoteToTopLevelIfNeeded(PackageArchiveReader reader, NuGetProject
             foreach (var package in upgradeDependencyItems)
             {
-                InitPackageUpgradeIssues(folderNuGetProject, package, framework);
+                InitPackageUpgradeIssues(((MSBuildNuGetProject)Project).FolderNuGetProject, package);
```
get the `FolderNuGetProject` once outside the loop.

nkolev92 replied:
good point.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/NuGetProjectUpgradeWindowModel.cs:
```diff
@@ -158,29 +157,13 @@ private void PromoteToTopLevelIfNeeded(PackageArchiveReader reader, NuGetProject
             foreach (var package in upgradeDependencyItems)
             {
-                InitPackageUpgradeIssues(folderNuGetProject, package, framework);
+                InitPackageUpgradeIssues(((MSBuildNuGetProject)Project).FolderNuGetProject, package);
```
Does this class only work with `FolderNuGetProject`? Why not change the type of `Project`, so no typecasting is needed?

nkolev92 replied:
The whole thing is based around packages.config projects, will do another pass see what can be improved.

nkolev92 replied:
Change the type to MSBUildNuGetProject

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,123 @@
+        /// <param name="packageIdentities">A collection of <see cref="PackageIdentity"/> to get info for.</param>
+        /// <param name="nuGetFramework">framework for determining the dependency groups of packages</param>
+        /// <param name="dependencyInfoResource">The resource to fetch dependency info from. Could be http/file feed/global packages folder/solution packages folder.</param>
+        /// <param name="sourceCacheContext">Caching context. Only reallly applicable when the dependency info resource is http based</param>
```
`reallly`?

nkolev92 replied:
realllly! :D 
Nice catch.

nkolev92 replied:
Fixed. I added 2 more l's

zivkan reviewed (commented):
I'm not sure if the goal of the PR/issue is to simply make it possible to get information, or if it should also make it easy. I don't know the NuGet SDK very well at all. I assume that `GetPackagesWithDependants` is the main method needed to get the package heirarchy from a packages.config file, but I'm not sure how easy/obvious it is for someone to get a `IList<PackageDependencyInfo>` from a packages.config file.

kartheekp-ms commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,123 @@
+            }
+            catch (NuGetProtocolException)
+            {
+                return null;
```
just curious, don't we need to log the error message before returning null ?

nkolev92 replied:
Good point, we can log it. 
Originally my thought was, let's not change the experience in the migrator, but this would be a good improvement.

kartheekp-ms commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,123 @@
+            return packageWithDependants;
+        }
+
+        private class PackageWithDependants : IPackageWithDependants
```
This class is also declared in the `PackageGraphAnalysisUtilitiesTests.cs`. Can we avoid duplication ?

nkolev92 replied:
I can make it internal. 
I don't really want to make this externally available (read public) as it's an extremely primitive and even unusable model :)

nkolev92 commented:
@kartheekp-ms @zivkan Addressed your feedback. 

> I'm not sure if the goal of the PR/issue is to simply make it possible to get information, or if it should also make it easy. 

A little bit of both, but primarily the first. 

> I don't know the NuGet SDK very well at all. I assume that GetPackagesWithDependants is the main method needed to get the package heirarchy from a packages.config file, but I'm not sure how easy/obvious it is for someone to get a IList<PackageDependencyInfo> from a packages.config file.

I refactored this out in such a way that the fact there's a project is irrelevant (see the Utilities class has no mention of project). 

If you have any specific suggestions, I'd be happy to discuss.

nkolev92 committed:
address feedback

nkolev92 committed:
add missing file

zivkan commented on src/NuGet.Clients/NuGet.VisualStudio.Client/.vsixignore:
```diff
@@ -86,4 +86,5 @@ Microsoft.VisualStudio.Workspace.VSIntegration.Contracts.dll
-VSLangProj157.dll
\ No newline at end of file
+VSLangProj157.dll
+System.Resources.Extensions.dll
```
The rest of the file appears to be sorted alphabetically, so this seems out of place 😛

donnie-msft replied:
First of all, are you really intending on checking in the ignore file?

Secondly, not sure alphabetical is a big deal to me, but would be nice to separate it with a comment explaining why you put it there. I think the reason is because "some change happened that we are unaware of but it breaks our compilation sometimes".

nkolev92 replied:
![image](https://user-images.githubusercontent.com/2878341/64813665-b7427380-d556-11e9-9ccf-4d29e6b8d9ef.png)

It's required by the project system assembly, so it's not our job to bring it into the product. 

I don't think we should put comments on every exclusion. 
For most of them the answer is, an internal component needs it, and we are not self contained.

I did sort it alphabetically. 

Originally I did not intend to check this in, but now we know why that's happening.

zivkan replied:
> now we know why that's happening

Any idea why it's not consistent? That's what I find weird. If it was every time, fine, but I don't understand why sometimes it builds fine without the vsixignore change.

nkolev92 replied:
🤷‍♂

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,124 @@
+        /// This method assumes that there is a 1-to-1 relationship between <paramref name="packageDependencyInfos"/> and <paramref name="packagesWithDependants"/>. 
+        /// </summary>
+        /// <remarks>This method will not handle null <paramref name="packageDependencyInfos"/> and <paramref name="packagesWithDependants"/></remarks>
+        public static void PopulateDependants(IList<PackageDependencyInfo> packageDependencyInfos, IEnumerable<IPackageWithDependants> packagesWithDependants)
```
From an API design point of view, I don't like this. I prefer pure functions that have no side effects. How about making the return type IEnumerable<PackageDependencyInfo>, change the `.Add` in the method to `yield return`, and let the caller use `.AddRange` if it really wants to mutate an existing list, rather than creating a new list?

nkolev92 replied:
I don't necessarily mind mutating method as long as they call it out. 
We shouldn't need the extra allocations if avoidable. 

Anyway, I like your ideas in this and the below comment, I'll work through them and re-request once I'm happy with what I have. So wait to reply to this until I push my changes :D

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,124 @@
+        /// Given <paramref name="packageDependencyInfos"/> generates a collection of <see cref="IPackageWithDependants"/> with the dependants populated correctly.
+        /// </summary>
+        /// <returns>A collection of <see cref="IPackageWithDependants"/></returns>
+        public static IEnumerable<IPackageWithDependants> GetPackagesWithDependants(IList<PackageDependencyInfo> packageDependencyInfos)
```
this method appears to mutate the input, in addition to returning a value. Neither the name of the method, nor the xmldoc suggest that the input will be changed.

nkolev92 replied:
It doesn't modify the input.

zivkan commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Models/NuGetProjectUpgradeWindowModel.cs:
```diff
@@ -94,11 +92,12 @@ public ObservableCollection<NuGetProjectUpgradeDependencyItem> UpgradeDependency
 
         public IEnumerable<NuGetProjectUpgradeDependencyItem> DirectDependencies => UpgradeDependencyItems
-                .Where(upgradeDependencyItem => !upgradeDependencyItem.DependingPackages.Any());
+                .Where(upgradeDependencyItem => PackageGraphAnalysisUtilities.TopLevelPackagesPredicate(upgradeDependencyItem));
```
I don't know if it has a performance impact by avoiding an additional anonymous function, but you can use `.Where(PackageGraphAnalysisUtilities.TopLevelPackagesPredicate);`

nkolev92 replied:
It's not necessary, so I'll remove it. 
It's an extra allocation, it doesn't make a huge difference, but it's a good practice to avoid extra work especially when it arguably makes things more readable.

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/IPackageWithDependants.cs:
```diff
@@ -0,0 +1,15 @@
+
+namespace NuGet.PackageManagement
+{
+    public interface IPackageWithDependants
```
It appears this is an interface, rather than a class, only because it's being refactored from `NuGetProjectUpgradeDependencyItem`, but it feels weird from an API point of view. Models are usually classes, not interfaces.

In particular, `PackageGraphAnalysisUtilities.GetPackagesWithDependants` creates its own internal instances, so they'll be incompatible with `NuGet.PackageManagement.UI`'s `NuGetProjectUpgradeDependencyItem`. Anyone working on the UI should probably know that they shouldn't call `GetPackagesWithDependants` without converting the objects to what would otherwise appear to be a compatible type.

nkolev92 replied:
> In particular, PackageGraphAnalysisUtilities.GetPackagesWithDependants creates its own internal instances, so they'll be incompatible with NuGet.PackageManagement.UI's NuGetProjectUpgradeDependencyItem

It's not being called there. 
There was feedback that we should have a method that can generate the toplevel and transitive packages without requiring much work from the caller.

zivkan commented on test/NuGet.Core.Tests/NuGet.PackageManagement.Test/Utility/PackageGraphAnalysisUtilitiesTests.cs:
```diff
@@ -0,0 +1,270 @@
+        // Installed list: A 1.0.0, B 1.0.0, C 1.1.0, D 1.0.0
+        // The scenario is a bump in the version of C
+        [Fact]
+        public void PackageGraphAnalysisUtilities_PopulateDependants_Succeeds()
```
`PackageGraphAnalysisUtilities_PopulateDependants_Succeeds` and `PackageGraphAnalysisUtilities_GetPackagesWithDependants_Succeeds` are almost identical, which is a code smell suggesting that the `PopulateDependants` and `GetPackagesWithDependants` APIs are not well designed.

nkolev92 replied:
why?
PackageGraphAnalysisUtilities_GetPackagesWithDependants was feedback to provide an extra method to avoid requiring an implementation of an interface for consumers of these utilities. 
It's ok if they call each other.

zivkan replied:
It suggests that the two methods being tested have the same implementation and therefore violates the Don't Repeat Yourself principal. It'd be more understandable if the two methods were an overload of the same name, but different names suggest they should do different things.

nkolev92 replied:
> It suggests that the two methods being tested have the same implementation and therefore violates the Don't Repeat Yourself principal.

I disagree. You can't know that. You can't make the assumption that they don't call each other.

>  It'd be more understandable if the two methods were an overload of the same name, but different names suggest they should do different things.

They do, one returns, the other one populates a preexisting model.

nkolev92 committed:
initial feedback

nkolev92 committed:
address feedback

nkolev92 commented:
Synced up offline. 
Incorporated some suggestions by @zivkan and I'm happier with the API right now. 
Ready for review again.

kartheekp-ms commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,103 @@
+
+            foreach(var package in packageDependencyInfos)
+            {
+                packageWithDependants.Add(new PackageWithDependants(package, GetDependantPackages(package, packageDependencyInfos).ToArray()));
```
Do we need to invoke ToArray() as it increases time complexity ?

zivkan replied:
`GetDependantPackages` returns `IEnumerable<T>`, but `PackageWithDependants` constructor accepts `IList<T>` (although I propose it should be `IReadOnlyList<T>`). Either `ToList` or `ToArray` is needed because of type checking, and I don't think it's appropriate for `PackageWithDependants` to accept an IEnumerable instead.

nkolev92 replied:
It doesn't increase the time complexity. 
It would've been enumerated at some point anyways. 

However, it does add an extra allocation.

The class takes an ILIst though so it can change to a list. 

edit: I meant to write this before Andy reviewed it, but went to a meeting. 
I agree with the feedback, I'll change it to list.

nkolev92 replied:
So I can change to IList, but I can't remove the toarray. 

The reason for that is that you can't infer a less specific collection from a more specific one. 

By changing PackageWithDependants to accept an IReadOnlyList, there are no conversions wasted.

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/PackageWithDependants.cs:
```diff
@@ -0,0 +1,25 @@
+    {
+        public PackageIdentity Identity { get; }
+
+        public IList<PackageIdentity> DependantPackages { get; }
```
```suggestion
        public IReadOnlyList<PackageIdentity> DependantPackages { get; }
```

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,103 @@
+        /// For example, the feed is not accessible.
+        /// </summary>
+        /// <param name="packageIdentities">A collection of <see cref="PackageIdentity"/> to get info for.</param>
+        /// <param name="nuGetFramework">framework for determining the dependency groups of packages</param>
```
```suggestion
        /// <param name="nuGetFramework">Framework for determining the dependency groups of packages</param>
```

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,103 @@
+        /// <param name="packageIdentities">A collection of <see cref="PackageIdentity"/> to get info for.</param>
+        /// <param name="nuGetFramework">framework for determining the dependency groups of packages</param>
+        /// <param name="dependencyInfoResource">The resource to fetch dependency info from. Could be http/file feed/global packages folder/solution packages folder.</param>
+        /// <param name="sourceCacheContext">Caching context. Only really applicable when the dependency info resource is http based</param>
```
```suggestion
        /// <param name="sourceCacheContext">Caching context. Only really applicable when the dependency info resource is http based.</param>
```

zivkan commented on src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        /// <summary>
+        /// Given a package(id and version) and a list of other package, it returns all the packages that have a reference to the exact ID and version.
```
```suggestion
        /// Return the packages from a list that have a dependency on a specific package (id and version).
```

zivkan reviewed (approved):
i ❤️ nitpicking.

nkolev92 committed:
Update src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs

Co-Authored-By: Andy Zivkovic <zivkan@users.noreply.github.com>

nkolev92 committed:
Update src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs

Co-Authored-By: Andy Zivkovic <zivkan@users.noreply.github.com>

nkolev92 committed:
Update src/NuGet.Core/NuGet.PackageManagement/Utility/PackageGraphAnalysisUtilities.cs

Co-Authored-By: Andy Zivkovic <zivkan@users.noreply.github.com>

nkolev92 committed:
work

nkolev92 committed:
Merge branch 'dev-nkolev92-topLevelPackageUtility' of https://github.com/NuGet/NuGet.Client into dev-nkolev92-topLevelPackageUtility

nkolev92 committed:
fix compiel error
